### ISSUE TITLE
Flash of unstyled content

### ISSUE THREAD

**Issue by programcsharp** [MiniProfiler/dotnet#364]
I'm getting a flash of unstyled content, particularly on Firefox:

![image](https://user-images.githubusercontent.com/994270/53016288-822fd180-340a-11e9-80b2-682f009d0829.png)

Skimming the code, it looks like the CSS is deferred which means there's a moment when the popup is rendered on the page but doesn't have style: https://github.com/MiniProfiler/dotnet/blob/6b3a42a65fed15b8f8b35804668fbc2ea429402b/src/MiniProfiler.Shared/ui/lib/MiniProfiler.ts#L350

An idea for a fix would be to stick in some default style in doInit() to hide the wrapper initially (before the deferInit):

 $('head').append(`<style type="text/css">.mp-results { display: none }</style>`);

And then in the includes.less, show the wrapper (since the style will be in place so it's hidden correctly by the other styles):

.mp-results { display: block }

If that solve is sufficient, let me know and I'll submit a PR.

Another thought would be to insert the css as a string into the .js -- gzipped it'd only add ~3KB to an already 47KB file and saving a network request is probably worth it since it's cached anyway... oh wait, it isn't. Is there a reason ProcessRequest doesn't slap a far futures expiry on there? It's got the version number in the URL so it should be a prime candidate.

**Comment by NickCraver** [comment]
> Is there a reason ProcessRequest doesn't slap a far futures expiry on there?

Nope...this used to be the case due to lack of proper cache breakers long ago - now we have per-version breakers that are fine to cache for say a month. Overall, it's a regression for sure. It should be cached which makes this a 1-time load thing and not worth complicating further. I'll poke this issue when a new build is up with some cache headers.

**Comment by programcsharp** [comment]
Couldn't that still lead to a flash? Looks like it's doing a 100ms setInterval on deferInit, so you could have content in place before the CSS, even with cache. Doing the initial CSS hide + show on real CSS load would ensure that doesn't happen.

**Comment by NickCraver** [comment]
Could it happen the very first time a user ever loaded MiniProfiler (usually via the corner item, which isn't shown until the JS loads later)? Yes. But in practice, you won't notice that...because it happens before the first corner block shows, and then it's not an issue for a month.

**Comment by programcsharp** [comment]
Well, the way I read:

https://github.com/MiniProfiler/dotnet/blob/6b3a42a65fed15b8f8b35804668fbc2ea429402b/src/MiniProfiler.Shared/ui/lib/MiniProfiler.ts#L257

is that it's designed to be able to wait on some other thing to be loaded... meaning that if anybody is actually using that the unstyled content would remain until the condition is satisfied.

**Comment by NickCraver** [comment]
A fix for the cache headers is in 04f8ac465349a85a77d9fa4c6528fb75e63e157c and will be available on MyGet shortly. I'll give it a proper spin on Stack Oveflow next week before promoting to NuGet, it's available for all in [the MyGet feed](https://www.myget.org/gallery/miniprofiler) as `4.0.165` before then :)

Context on the other proposals:
An initial CSS hide would then depend on JS to show it. It turns what can be instant today into a flicker instead. It breaks user scripts, breaks the option to hide/show it by default (Alt+P is a keyboard shortcut) and introduces races and complexity. IMO, that's trying to solve a 1% problem at the expense of the 99% case and adding more/compounding problems. Let's go as simple as possible first, and only make more complicated changes if they're warranted.

**Comment by programcsharp** [comment]
Upgraded to 4.0.165 and it's definitely caching now but it doesn't help at all. Still a big fat flash of unstyled content in Firefox. Have you tried FF? It much rarer and split second in Chrome but FF seems to do the render cycle differently and leave that thick table right there on screen -- basically deferring the <link> css parse until after the JS on the page has executed. Look below for a gif recording... and when you can see it in a 10fps recording you know it's pretty bad 😁 

I'm not understanding your reticence to take the easy win and drop in CSS. You don't need any more JS, just drop in a display: none first in a <style> tag and then a display: block in the CSS file.

#367 is a two liner fix for this. Thoughts? Only other idea I have is a MiniProfiler.RenderCss that can be done in the <head> on render so the CSS is parsed in normal order by Firefox.

![flash of unstyled content](https://user-images.githubusercontent.com/994270/54065960-15c41900-41e5-11e9-8509-43a316befcad.gif)

**Comment by NickCraver** [comment]
I'll try and take a look at this next week: the reluctance is this will break people's layouts (including Stack Overflow's). Just because a div is a div doesn't mean it's `display: block` after hidden. Unfortunately, that's a bad assumption with flexbox in the wild. It's non-trivial to detect what should happen there, and jQuery does the wrong thing too.

Firefox shouldn't be doing this - I'll try and repro locally when back home to figure out what's happening there. The CSS load should be instantaneous. I want to avoid going crazy with flexbox detection unless we *have* to, but need a repro locally to access that.

But! I have a thought: I wonder though, if this is a threading escape issue in Firefox.

Can you try instead of the PR fix, *only* replacing `doInit();` with `setTimeout(doInit, 1);` on line 352? Let's see if that makes Firefox evaluate the style addition and load before creating the objects. If that works, it'll solve the problem without poking the flexbox beast :)

**Comment by NickCraver** [comment]
@programcsharp Were you able to test the simpler fix with a timeout here?
